File: lexicon.py
import asyncio
from json import loads
from utils import get_url_contents

lexicon = {}

async def update_lexicon():
    global lexicon
    
    data = await get_url_contents("https://mr-martian.github.io/Sajem_Tan/lexicon.js")
    data = data.split("var lexicon = ")[1].split("};")[0] + "}"
    data = "\n".join([i for i in data.split("\n") if "//" not in i])

    lexicon = loads(data)

def format_definition(ls):
    result = []
    for definition in ls:
        if isinstance(definition, str):
            result.append(definition)
        elif isinstance(definition, list):
            hint = definition[0]
            if hint == ("archaic" or "see also"):
                result.append(f"*{hint}:* {definition[1]}")
            elif hint == "example":
                result.append(f"*example:* {definition[1]} ({definition[2]})")
            elif hint == "suf-example":
                result.append(f"*example:* {definition[1]} \u2192 {definition[2]}")
    return "; ".join(result)

def define(word):
    global lexicon

    result = []
    for catname, category in lexicon.items():
        for entryname, entry in category.items():
            if isinstance(entry, list):
                if entryname == word:
                    result.append(f"**({catname})** {format_definition(entry)}")
            elif isinstance(entry, dict):
                for subentryname, subentry in entry.items():
                    if subentryname == word:
                        result.append(
                            f"**({catname} > {entryname})** {format_definition(subentry)}"
                        )
    return "__**{}**__\n\n{}".format(
        word, "\n".join(result)
    )

File: lexicon.rb
require 'net/http'
require 'json'

class Lexicon 
    def initialize
        @raw = JSON.parse(
            Net::HTTP.get('sajemtan.github.io', '/lexicon.js')
                .split("var lexicon =")[1]
                .split("};")[0] +
                "}"
        )
    end

    def define(str)
        result = []
        @raw.each { |catname, category|
            category.each { |entryname, entry|
                if entry.class == Array and entryname == str
                    result << [catname, entry]
                    # "**(#{catname})** #{format_def(entry)}"
                elsif entry.class == Hash
                    entry.each { |subentname, subentry|
                        if subentname == str
                            result << ["#{catname} > #{entryname}", subentry]
                            # "**(#{catname} > #{entryname})** #{format_def(subentry)}"
                        end
                    }
                end
            }
        }

        result

        # if result.length > 0
        #     "__**#{str}**__\n\n#{result.join "\n"}"
        # else
        #     "No results found for #{str}"
        # end
    end

    def define_formatted(str)
        format_entry_array(str, define(str))
    end
end

def format_entry(pos, data)
    result = []

    data.each { |e|
        if e.class == String
            result << e
        elsif e.class == Array
            case e[0]
                when "archaic"
                    result << "*archaic, see:* #{e[1]}"
                when "see also"
                    result << "*see also:* **#{e[1]}**"
                when "example"
                    result << "*example:* #{e[1]} (#{e[2]})"
                when "suf-example"
                    result << "*example:* #{e[1]} \u2192 #{e[2]}"
            end
        end
    }

    "**(#{pos})** #{result.join "; "}"
end

def format_entry_array(str, data)
    if data.empty?
        "No results found for #{str}"
    else
        "__**#{str}**__\n\n#{data.map {|e| format_entry(*e)}.join "\n"}"
    end
end
